sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -194,18 +197,145 @@ def display_data(self):
-      if size is None or size <= 0:
-        raise ValueError(
-            'Collection %s not found or average doc size is '
-            'incorrect', self.coll)
-      return size
-
-  def _get_document_count(self):
+      name_space = '%s.%s' % (self.db, self.coll)
+      return (client[self.db].command(
+          'splitVector',
+          name_space,
+          keyPattern={'_id': 1},
+          min={'_id': start_pos},
+          max={'_id': end_pos},
+          maxChunkSize=desired_chunk_size_in_mb)['splitKeys'])
+
+  def _merge_id_filter(self, range_tracker):
+    # Merge the default filter with refined _id field range of range_tracker.
+    all_filters = self.filter.copy()
+
+    # if there are no additional filters, initialize empty additional filters,
+    # see more at https://docs.mongodb.com/manual/reference/operator/query/and/
+    if '$and' not in all_filters:
+      all_filters['$and'] = []
+
+    # add additional range filter to query. $gte specifies start position (
+    # inclusive) and $lt specifies the end position (exclusive), see more at
+    # https://docs.mongodb.com/manual/reference/operator/query/gte/ and
+    # https://docs.mongodb.com/manual/reference/operator/query/lt/
+    all_filters['$and'] += [{
+        '_id': {
+            '$gte': range_tracker.start_position(),
+            '$lt': range_tracker.stop_position()
+        }
+    }]
+    return all_filters
+
+  def _get_head_document_id(self, sort_order):
     with MongoClient(self.uri, **self.spec) as client:
-      return max(client[self.db][self.coll].count_documents(self.filter), 0)
+      cursor = client[self.db][self.coll].find(filter={}, projection=[]).sort([
+          ('_id', sort_order)
+      ]).limit(1)
+      try:
+        return cursor[0]['_id']
+      except IndexError:
+        raise ValueError('Empty Mongodb collection')
+
+  def _replace_none_positions(self, start_position, stop_position):
+    if start_position is None:
+      start_position = self._get_head_document_id(ASCENDING)
+    if stop_position is None:
+      last_doc_id = self._get_head_document_id(DESCENDING)
+      # increment last doc id binary value by 1 to make sure the last document
+      # is not excluded
+      stop_position = _ObjectIdHelper.increment_id(last_doc_id, 1)
+    return start_position, stop_position
+
+
+class _ObjectIdHelper(object):
+  """A Utility class to manipulate bson object ids."""
+
+  @classmethod
+  def id_to_int(cls, id):
+    """
+    Args:
+      id: ObjectId required for each MongoDB document _id field.
+
+    Returns: Converted integer value of ObjectId's 12 bytes binary value.
+
+    """
+    # converts object id binary to integer
+    # id object is bytes type with size of 12
+    ints = struct.unpack('>III', id.binary)
+    return (ints[0] << 64) + (ints[1] << 32) + ints[2]
+
+  @classmethod
+  def int_to_id(cls, number):
+    """
+    Args:
+      number(int): The integer value to be used to convert to ObjectId.
```

Comment:
Do all versions of Python that this code can run on use arbitrary-precision integers by default?

Reply:
precommit tests covers python2.7, 3.5, 3.6, 3.7